### PR TITLE
As an OmniNav user, I want the new OmniNav design implemented in Dev Cascade to be fully dynamic.

### PR DESCRIPTION
**Pull request by tatwell**
Trello Story: https://trello.com/c/JiDzj9Lv

This story addresses part of this issue: #239

### CODE REVIEW

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 1]
```diff
@@ -0,0 +1,812 @@
...
+ * OmniNav v2 format.
+ * Chapman.edu/_components/omni_nav_v2/omni_nav.vtl
+ *
+ * It would make sense to move the various component-specific macros (like those
+ * for Primary Nav) into their own file and import them like so:
+ *
+ * #import( "/_components/omni_nav_v2/macros/primary_nav.vm" )
+ *
+ * But this makes debugging the code in Cascade's format editor much harder. It is
+ * simpler to just keep it all in a well-structured single file.
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: I would love to make this more modular but found it impractical for development.

I added a Velocity Best Practices wiki page:

- https://wimops.chapman.edu/wiki/Velocity

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 2]
```diff
@@ -0,0 +1,812 @@
...
+  [ 'academics', 'Academics', 'default', '/' ],
+  [ 'admission', 'Admission', 'default', '/' ],
+  [ 'alumni', 'Alumni', 'default', '/' ],
+  [ 'campus-services', 'Campus Services', 'default', '/' ],
+  [ 'faculty-staff', 'Faculty and Staff', 'default', '/' ],
+  [ 'families', 'Families', 'default', '/' ],
+  [ 'research', 'Research', 'default', '/' ],
+  [ 'students', 'Students', 'default', '/' ],
+  [ 'support-chapman', 'Support Chapman', 'default', '/' ],
+  [ 'communication', 'SCHOOL OF COMM', 'soc', '/communication/' ]
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: I'm still waiting for Ross to get back to me on this list. He's reviewing the short names (2nd item in each list).

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 3]
```diff
@@ -0,0 +1,812 @@
...
+])
+
+##
+## Vars
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: Next two sections is where the "data" gets set.

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 4]
```diff
@@ -0,0 +1,812 @@
...
+      #set ( $discard = $umbrellaAssets.add($childAsset) )
+    #end
+  #end
+#end
+
+
+##
+## Template Macros
+##
+#macro ( buildOmniNav )
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: Main macro that builds the OmniNav. Macros for each part of OmniNav are organized into sections below.

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+                <li><a class="icon icon-pencil5" href="/admission/graduate/applynow.aspx">Graduate Application</a></li>
+                <li><a class="icon icon-calculate2" href="/admission/undergraduate/afford.aspx">Affordability</a></li>
+                <li><a class="icon icon-calculate" href="/students/tuition-and-aid/financial-aid/net-cost-calculator/">Financial Aid Calculator</a></li>
+                <li><a class="icon icon-office" href="/admission/undergraduate/visit.aspx">Campus Tours</a></li>
+              </ul>
+            </div>
+          </li>
+          <li class="primary-link">
+            <a>Faculty</a>
+            <div class="global-nav-dropdown">
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: Notice: next 2 sections are all Law links and window icons. Let me know if this should be amended. Does not necessarily have to be changed now. Can be change in "final touches" story.

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 6]
```diff
@@ -0,0 +1,812 @@
...
+        #buildSecondaryMenuItem( $asset )
+      #end
+    </ul>
+  </div>
+#end
+
+#macro ( buildSecondaryMenuItem $asset )
+  ## This builds a menu link for the secondary menu. It will either be a link or a
+  ## drop-down menu with multiple links.
+  #set ( $metaHideSetting = $asset.metadata.getDynamicField('Hide from navigation').value )
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: Notice how metadata is accessed here using Cascade Adapter API.

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+  #set ( $isVisible = $metaHideSetting != 'Yes' )
+  #set ( $assetType = $asset.assetType )
+
+  ## If hidden, continue. But velocity doesn't have a continue operator:
+  ## https://stackoverflow.com/questions/7530019/apache-velocity-foreach-loop-continue
+  #if ( $isVisible )
+    #if ( $assetType == 'page' )
+      <li><a href="${asset.path}">${_EscapeTool.xml($asset.metadata.displayName)}</a></li>
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: I would not expect `_EscapeTool` to be required here. But it's hard to tell with the Adapter API methods. And error messages when it is needed are not clear.

File: .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+    <!-- $asset.name : $asset.path ($asset.children.size()) -->
+  #end
+#end
+
+
+##
+## Main OmniNav HTML Block
+##
+#setVars()
+#buildOmniNav()
```
Comment (tatwell) [inline .cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav.vtl]: Here's the main block. I do like this setVars, buildHTML pattern.

File: app/assets/javascripts/omni-nav-v2.js [hunk 9]
```diff
@@ -105,9 +106,21 @@ var OmniNav2 = (function() {
...
+    var enableMenusToggle = function() {
+      // Enables toggle to slide main/umbrella menus back and forth.
+      $('a.toggle-menu-label').on('click', function(e) {
+        console.debug('toggle menus now!');
+        // Toggles headers.
+        $('div#umbrella-logo').toggle('blind');
+
+        // Slide-toggles the menus.
+        $('div#off-canvas-umbrella').toggle('slide');
+      });
```
Comment (tatwell) [inline app/assets/javascripts/omni-nav-v2.js]: Notice I'm using jQueryUI API here. (At least, I think I am.) Old OmniNav wasn't. It was a mess.

File: app/assets/stylesheets/omni_nav/omni_nav_v2/omni-nav.scss [hunk 10]
```diff
@@ -30,7 +30,36 @@ html.omni-nav-v2 {
   width: 100%;
   top: 0;
   font-family: "futura-pt", verdana, sans-serif;
-  z-index: 666;
+
+  // Needs to be high enough to put off-canvas-nav on top of anything else on page.
+  z-index: 1000;
```
Comment (tatwell) [inline app/assets/stylesheets/omni_nav/omni_nav_v2/omni-nav.scss]: This fixed that issue where OffCanvasNav was behind a block on one-column page.

**Comment by tatwell** [comment]
Final call for reviews on this PR. 

I need to merge this PR to wrap up this story before today's demo and the end of this sprint.

Nothing is getting released with this story so any issues lurking here can be followed up on in the new OmniNav stories that follow this. But I strongly recommend reviewing the velocity format as it exemplifies the approach to formats I would like us to follow going forward.

**Comment by kaylaziegler** [comment]
Are we able to test this new omninav format with other pages in Dev-Cascade?

**Comment by tatwell** [comment]
Yes, we can test with other pages. To update a page:

- In Cascade, go to page, e.g. [Law Index](https://dev-cascade.chapman.edu/entity/open.act?id=8d4c2096c04d744c4cce810723448f72&type=page)
- Edit Page > Configure
- Under Regions > OMNI-NAV: replace format with  new `omni_nav.vtl` format and check No Block box.
- Click Save & Preview button (upper right)
- Click Submit to save updates
- Click Publish

Result:

- https://dev-www.chapman.edu/law/

**Comment by kaylaziegler** [comment]
Okay, I tested it on Academics and About, and there are some random pages/folders missing.

For example, Academics is missing Course Catalogs and Libraries but neither the pages or the folders are marked as hidden from navigation. Do you know why they wouldn't be showing up? https://dev-www.chapman.edu/academics/index.aspx

**Comment by tatwell** [comment]
@kaylaziegler Good question. Comparing these two pages, I see the same thing:

- New OmniNav: https://dev-www.chapman.edu/academics/index.aspx
- Old OmniNav: https://dev-www.chapman.edu/academics/academic-calendar.aspx

All I can assume is it's something unexpected with Adapter API object. 

I also noticed that the Umbrella Menu logo is missing.

My plan is to open a new story to address these issues and still merge this branch.

File: app/assets/javascripts/omni-nav-v2.js [hunk 11]
```diff
@@ -105,9 +106,21 @@ var OmniNav2 = (function() {
...
     var syncLinkWidths = function() {
       var width = $('#js-off-canvas-nav > ul').width();
-      offCanvasLinks.css('width', width);
+      $offCanvasLinks.css('width', width);
     };
 
+    var enableMenusToggle = function() {
+      // Enables toggle to slide main/umbrella menus back and forth.
+      $('a.toggle-menu-label').on('click', function(e) {
+        console.debug('toggle menus now!');
```
Comment (mirabelrice) [inline app/assets/javascripts/omni-nav-v2.js]: Do we have a standard for committing debug outputs?

**Comment by tatwell** [comment]
New story created for missing menu item issue identified by Kayla and missing logo issue I noticed:

- https://trello.com/c/NWW6Ttr8

File: app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss [hunk 12]
```diff
@@ -14,8 +14,10 @@
   }
 
   .off-canvas-overlay {
+    // This value is relative to OmniNav parent.
+    z-index: 1002;
```
Comment (mirabelrice) [inline app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss]: We have always started our `z-index` values from `666`, are we changing this now? Our other apps use this basis as well.

File: app/assets/javascripts/omni-nav-v2.js (hunk 11 above)
Comment (tatwell) [reply app/assets/javascripts/omni-nav-v2.js]: Yes, I'll remove that.

File: app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss (hunk 12 above)
Comment (tatwell) [reply app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss]: When `z-index` was at `666` we were getting that issue where OffCampus menu was underneath another page element.

It's possible that this was being applied in 2 different places, so this isn't necessarily a violation of that policy.

Also, I wasn't aware that this was our practice. I mean as much as I love

maxresdefault

and accept `666` as an intrinsically magic number, I'm not sure how best to standardize this.

I wonder if there are best practices on the use of z-index.

Comment (mirabelrice) [reply app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss]: If this works, its fine for now! I agree we should look into best practices, and set our standard.

Comment (tatwell) [reply app/assets/stylesheets/omni_nav/omni_nav_v2/off-canvas-nav.scss]: I'll add an action item for this.

**Pull request merged**